dgisim/src/dices.py

```diff
@@ -241,6 +272,206 @@ def just_satisfy(self, requirement: AbstractDices) -> bool:
+        return BIG_INT if no possibility
+        """
+        if element_supply + omni_supply >= element_needed + omni_needed:
+            if element_prioritized:
+                return element_needed + omni_needed
+            else:
+                return max((element_needed + omni_needed) - element_supply, 0)
+        else:
+            # return None
+            return BIG_INT
+
+    def smart_selection(
+            self,
+            requirement: AbstractDices,
+            game_state: Optional[GameState] = None,
+            local_precedence_arg: Optional[list[set[Element]]] = None,
+    ) -> Optional[ActualDices]:
+        """ trying to fill requirement with self._dices """
+
+        # result in dict format
+        result_dct: dict[Element, int] = defaultdict(int)
+
+        # optimisation check - if required dices > avaiable dices, break
+        if requirement.num_dices() > self.num_dices():
+            return None
+
+        supply: dict[Element, int] = defaultdict(int)
+        supply.update(self._dices)
+
+        need: dict[Element, int] = defaultdict(int)
+        need.update(requirement._dices)
+
+        local_precedence: list[set[Element]] = \
+            [] if local_precedence_arg is None else local_precedence_arg
+        local_precedence_reversed: list[set[Element]] = local_precedence[::-1]
+        assert local_precedence is not None
+        assert local_precedence_reversed is not None
+        assert len(local_precedence) <= self._NUMBER_OF_ELEMENTS + 2
+
+        # list of first priority elements
+        first_priority_elements: set[Element] = (set() if local_precedence == [] else local_precedence[0])
+
+        @cache
+        def get_local_priority(element: Element) -> tuple[int, int]:
+            """ get priority by element, lower tuple is bigger priority and should be spent first """
+            nonlocal local_precedence
+            if element == Element.OMNI:
+                return self._NUMBER_OF_ELEMENTS + 2, -1  # OMNI has biggest tuple
+            for i, elem_set in enumerate(local_precedence, start=1):
+                if element in elem_set:
+                    return (self._NUMBER_OF_ELEMENTS + 1 - i,
+                            self._ELEMENTS_AND_OMNI_BY_DECREASING_GLOBAL_PRIORITY_DCT[element]
+                    )
+            else:
+                return -1, self._ELEMENTS_AND_OMNI_BY_DECREASING_GLOBAL_PRIORITY_DCT[element]
+
+        def check_asserts():
+            """
+            assert non-negativity of variables:
+            need supply result_dct
+            """
+
+            def check_dict(name: str, dct: dict):
+                for el, val in dct.items():
+                    assert val >= 0, f"{el} in {name} < 0 {val=}"
+
+            nonlocal result_dct, supply, need
+
+            check_dict("result_dct", result_dct)
+            check_dict("supply", supply)
+            check_dict("need", need)
+
+        # 1st step - fill OMNI requirement
```

Comment:
you should probably swap the _first step_ with the _second step_; the reason is:

To fill the pure-elements, the solution is simple and straight forward, you fill each with the corresponding pure-elements and if there's not enough, fill the rest with OMNI.

Then without worrying about filling the pure-elements, the logic and reasoning to fill the OMNI will be a lot simpler.

Reply:
I don't know if I understand it correctly, but the problem with filling OMNI is first_priority elements, so my code tries to solve it in _first step_. It select which element to use, using data as how many dices does it spend from OMNI and first_priority pure elements (number of elements, if element is first_priority, 0 if no). It allows to minimize first_priority+omni use.

However, I don't know, if this order of _steps_ is necessary, it requires thinking. Write me your thoughts about it, I think I can miss some important points.